### PR TITLE
Refactor halo2 backend

### PR DESCRIPTION
**Pull request by leolara**

### CODE REVIEW

**Comment by alxkzmn** [comment]
@leolara the test is failing, could you please check? Also, there are changes to k calculation in my other pull request https://github.com/privacy-scaling-explorations/chiquito/pull/268, could we first merge them?

File: src/plonkish/backend/halo2.rs [hunk 1]
```diff
@@ -496,17 +549,76 @@ pub fn halo2_verify(
 
 /// Halo2 setup
 pub struct Setup {
+    pub k: u32,
```
Comment (alxkzmn) [inline src/plonkish/backend/halo2.rs:552]: The fields of `Setup` don't have to be public as you have exposed them through the getters.

File: src/plonkish/backend/halo2.rs [hunk 2]
```diff
@@ -559,142 +671,90 @@ fn assign_witness(
...
+// .map(|c| c.plonkish_ir.num_rows)
+// .max()
+// .unwrap_or(0);
+//
+// let (compiled, k) = self.compile_middleware(tallest_subcircuit_height).unwrap();
+// let setup = create_setup(rng, compiled, k);
+//
+// Halo2Prover::new(setup, self.clone())
+// }
+// }
```
Comment (alxkzmn) [inline src/plonkish/backend/halo2.rs:718]: Leftover code

**Comment by leolara** [comment]
Done both things

**Review by alxkzmn — APPROVED** [review]
Looks good! Should I merge?

**Pull request merged**